### ISSUE TITLE
Temporary Roles

### ISSUE THREAD

**Issue by DeepDeepDeep** [FivePixels/Tribble#22]
It would be nice if the bot had an option for temporary roles that would expire after a predefined time—Ex: 1 month/week.
This would be useful for monthly/weekly subscriptions. Maybe a reminder feature before their subscription expired would be a QoL feature.

**Comment by FivePixels** [comment]
We would need a way to store when a user was granted a role, and when to expire said role. I think we should iron out the design before implementing it, otherwise, it will be a beast to maintain. 

Some of my initial thoughts are on the startup of the bot, we check for expired roles. Anyone whose time has passed will get their role removed in a batch process. 

If we wanted to support temporary roles, we would need some way to continually poll for changes, which will probably get heavy after some time. We could ping the database for changes, say, every five seconds. 

Let me know your thoughts, but we should work out an implementation plan before just writing code. Otherwise, we're going to build a system that works one way and doesn't allow for easy changes, and will subsequently require a large refactor in the future.

**Comment by DeepDeepDeep** [comment]
> Some of my initial thoughts are on the startup of the bot, we check for expired roles. Anyone whose time has passed will get their role removed in a batch process.

Yep, that is how you implemented it before with epochs. You would check every minute and remove it if it was expired. 

`
function checkRoles() {
    paidAccess.fetchEverything();
    console.log('Checking paid access roles.')
    now = Math.floor(Date.now() / 1000);
    paidAccess.forEach((value, key) => {
        console.log(`user: ${key}, epoch: ${value}`)
        // now - value = seconds since epoch was stored, 60 seconds * minutes stored in env
        if (now - value > 60 * env.TEMPROLE_TIME) {
            // the user's access has expired
            guild.members.fetch(`${key}`).then(member => {
                try {
                    member.roles.remove(purchasedRole);
                    dateExpired = new Date((value + (60 *env.TEMPROLE_TIME)) * 1000)
                    upgradeLogChannel.send(`<@${member.id}> subscription expired on ${dateExpired.toISOString().replace(/T/, ' ').replace(/\..+/, '')}.`)
                    paidAccess.delete(key)
                    console.log(`${member.username} access expired. Role has been removed.`)
                } catch (error) {
                    console.log(error)
                }
            })
        }
    })
    console.log('Checking paid access complete.')
}
`
This didn't account for people leaving before their role expired so there had to be checks added.

`client.on('guildMemberRemove', async member => {
    paidAccess.fetchEverything();
    console.log('Checking paid access roles.')
    now = Math.floor(Date.now() / 1000);
    paidAccess.forEach((value, key) => {
        if(key == member.id)
        {
            dateExpired = new Date((value + (60 *env.TEMPROLE_TIME)) * 1000)
            console.log(`Member left was donor: ${member.user.username} expiration: ${dateExpired.toISOString().replace(/T/, ' ').replace(/\..+/, '')}.`)
            upgradeLogChannel.send(`<@${key}> left server. Expiration: ${dateExpired.toISOString().replace(/T/, ' ').replace(/\..+/, '')}.`)
            try {
            paidAccess.delete(key)
            console.log(`${member.username} Left server. Role has been removed.`)
            } catch (error) {
                console.log(error)
            } 
        }
    })
  });
`

**Comment by FivePixels** [comment]
I think before implementing this, we should implement a database for products. This way, a user can add products through a command and configure it through that, rather than the configuration file. This is talked about in #15, but I'm starting to see more potential behind storing this information in the database and making it easier for a user to configure products. The configuration as it stands is super complex for an end-user that doesn't know much about programming.
